gui/blueprint.lua

```diff
@@ -19,4 +19,160 @@ values set in the interface. See the `blueprint` documentation for information
+local gui = require('gui')
+local guidm = require('gui.dwarfmode')
+local widgets = require('gui.widgets')
+
+local function get_dims(pos1, pos2)
+    local width, height, depth = math.abs(pos1.x - pos2.x) + 1,
+            math.abs(pos1.y - pos2.y) + 1,
+            math.abs(pos1.z - pos2.z) + 1
+    return width, height, depth
+end
+
+ActionPanel = defclass(ActionPanel, widgets.Panel)
+ActionPanel.ATTRS{
+    get_mark_fn=DEFAULT_NIL,
+}
+function ActionPanel:init()
+    self:addviews{
+        widgets.Label{
+            text={{text=self:callback('get_action_text')}},
+            frame={t=0},
+        },
+        widgets.Label{
+            text='with the cursor.',
+            frame={t=1},
+        },
+    }
+end
+function ActionPanel:get_action_text()
+    if self.get_mark_fn() then
+        return 'Select the second corner'
+    end
+    return 'Select the first corner'
+end
+
+BlueprintUI = defclass(BlueprintUI, guidm.MenuOverlay)
+BlueprintUI.ATTRS {
+    presets={},
+    frame_inset=1,
+    focus_path='blueprint',
+}
+function BlueprintUI:init()
+    local summary = {
+        'Create quickfort blueprints\n',
+        'from a live game map.'
+    }
+
+    self:addviews{
+        widgets.Label{text='Blueprint', frame={t=0}},
+        widgets.Label{text=summary, text_pen=COLOR_GREY, frame={t=2}},
+        ActionPanel{get_mark_fn=function() return self.mark end, frame={t=5}},
+        widgets.Label{text={{text=function() return self:get_cancel_label() end,
+                             key='LEAVESCREEN', key_sep=': ',
+                             on_activate=function() self:on_cancel() end}},
+                             frame={t=8}},
+    }
+end
+
+function BlueprintUI:onAboutToShow()
+    if not dfhack.isMapLoaded() then
+        qerror('Please load a fortress map.')
+    end
+end
+
+function BlueprintUI:on_mark(pos)
+    self.mark = pos
+end
+
+function BlueprintUI:get_cancel_label()
+    if self.mark then
+        return 'Cancel selection'
+    end
+    return 'Back'
+end
+
+function BlueprintUI:on_cancel()
+    if self.mark then
+        self.mark = nil
+    else
+        self:dismiss()
+    end
+end
+
+-- Sorts and returns the given arguments.
+local function min_to_max(...)
+    local args = {...}
+    table.sort(args, function(a, b) return a < b end)
+    return table.unpack(args)
+end
+
+local fg, bg = COLOR_GREEN, COLOR_BLACK
+
+function BlueprintUI:onRenderBody()
+    if not self.mark then return end
+
+    local vp = self:getViewport()
+    local dc = gui.Painter.new(self.df_layout.map)
+
+    if gui.blink_visible(500) then
+        local cursor = df.global.cursor
+        -- clip blinking region to viewport
+        local _,y_start,y_end = min_to_max(self.mark.y, cursor.y, vp.y1, vp.y2)
+        local _,x_start,x_end = min_to_max(self.mark.x, cursor.x, vp.x1, vp.x2)
+        for y=y_start,y_end do for x=x_start,x_end do
+            local pos = xyz2pos(x, y, cursor.z)
+            if x == cursor.x and y == cursor.y then
```

Comment:
Perhaps this could use `same_xyz()` instead (unless performance is a concern here)? The check for z wouldn't affect anything since you've set `pos.z = cursor.z` on the line above. It would also allow removing the goto/continue statement without negating the compound statement here. For instance:

```lua
if not same_xyz(pos, cursor) then
   local stile = .......
end

Reply:
cleaner. i like. done : )